renterhost/registry/registry.go

```diff
@@ -0,0 +1,187 @@
+	// corresponds to the hash of a host's public key.
+	EntryTypePubKey
+)
+
+const (
+	// MaxValueDataSize is the maximum size of a Value's Data
+	// field.
+	MaxValueDataSize = 113
+)
+
+// A Value is stored in the host registry.
+type Value struct {
+	Tweak    types.Hash256
+	Data     []byte
+	Revision uint64
+	Type     uint8
+
+	PublicKey types.PublicKey
+	Signature types.Signature
+}
+
+// Key returns the key for the registry value.
+func (r *Value) Key() types.Hash256 {
+	return Key(r.PublicKey, r.Tweak)
+}
+
+// Hash returns the hash of the Value used for signing
+// the entry.
+func (r *Value) Hash() types.Hash256 {
+	h := types.NewHasher()
+
+	h.E.Write(r.Tweak[:])
+	h.E.WriteBytes(r.Data)
+	h.E.WriteUint64(r.Revision)
+	h.E.WriteUint64(uint64(r.Type))
+
+	return h.Sum()
+}
+
+// Work returns the work of a Value.
+func (r *Value) Work() types.Work {
+	var data []byte
+	switch r.Type {
+	case EntryTypePubKey:
+		// for public key entries the first 20 bytes represent the
+		// public key of the host, ignore it for work calculations.
+		data = r.Data[20:]
+	default:
+		data = r.Data
+	}
+
+	h := types.NewHasher()
+
+	h.E.Write(r.Tweak[:])
+	h.E.WriteBytes(data)
+	h.E.WriteUint64(r.Revision)
+
+	return types.WorkRequiredForHash(types.BlockID(h.Sum()))
+}
+
+// MaxLen returns the maximum length of an encoded Value. Implements
+// rpc.Object.
+func (r *Value) MaxLen() int {
+	return 32 + 8 + MaxValueDataSize + 8 + 1 + 32 + 64
+}
+
+// EncodeTo encodes a Value to an Encoder. Implements types.EncoderTo.
+func (r *Value) EncodeTo(e *types.Encoder) {
+	r.Tweak.EncodeTo(e)
+	e.WriteBytes(r.Data)
+	e.WriteUint64(r.Revision)
+	e.WriteUint8(r.Type)
+	r.PublicKey.EncodeTo(e)
+	r.Signature.EncodeTo(e)
+}
+
+// DecodeFrom decodes a Value from a Decoder. Implements types.DecoderFrom.
+func (r *Value) DecodeFrom(d *types.Decoder) {
+	r.Tweak.DecodeFrom(d)
+	r.Data = make([]byte, d.ReadPrefix())
+	d.Read(r.Data)
+	r.Revision = d.ReadUint64()
+	r.Type = d.ReadUint8()
+	r.PublicKey.DecodeFrom(d)
+	r.Signature.DecodeFrom(d)
+}
+
+// Key is the unique key for a Value.
+func Key(pub types.PublicKey, tweak types.Hash256) types.Hash256 {
+	// v1 compat registry key
+	// ed25519 specifier + LE uint64 pub key length + public key + tweak
+	buf := make([]byte, 16+8+32+32)
+	copy(buf, "ed25519")
+	binary.LittleEndian.PutUint64(buf[16:], 32)
+	copy(buf[24:], pub[:])
+	copy(buf[56:], tweak[:])
+	return types.HashBytes(buf)
+}
+
+// HostID returns the ID hash of the host for primary registry entries.
+func HostID(pub types.PublicKey) types.Hash256 {
+	// v1 compat host public key hash
+	// ed25519 specifier + LE uint64 pub key length + public key
+	buf := make([]byte, 16+8+32)
+	copy(buf, "ed25519")
+	binary.LittleEndian.PutUint64(buf[16:], 32)
+	copy(buf[24:], pub[:])
+	return types.HashBytes(buf)
+}
+
+// ValidEntry validates the fields of a registry entry.
+func ValidEntry(value Value) (err error) {
```

Comment:
my rule of thumb for naming here is:

- `ValidX` returns a `bool`
- `ValidateX` returns an `error`